emcc.py

```diff
@@ -1000,34 +1002,26 @@ def get_clang_command_asm():
 
   # Precompiled headers support
   if state.mode == Mode.PCH:
-    inputs = [i[1] for i in input_files]
-    for header in inputs:
-      if shared.suffix(header) not in HEADER_ENDINGS:
-        exit_with_error(f'cannot mix precompiled headers with non-header inputs: {inputs} : {header}')
-    cmd = get_clang_command() + inputs
-    if options.output_file:
-      cmd += ['-o', options.output_file]
+    cmd = get_clang_command() + newargs
     logger.debug(f"running (for precompiled headers): {cmd[0]} {' '.join(cmd[1:])}")
     shared.exec_process(cmd)
     assert False, 'exec_process does not return'
 
   if state.mode == Mode.COMPILE_ONLY:
-    inputs = [i[1] for i in input_files]
-    if all(get_file_suffix(i) in ASSEMBLY_ENDINGS for i in inputs):
-      cmd = get_clang_command_asm() + inputs
+    if options.output_file and get_file_suffix(options.output_file) == '.bc' and not settings.LTO and '-emit-llvm' not in state.orig_args:
+      diagnostics.warning('emcc', '.bc output file suffix used without -flto or -emit-llvm.  Consider using .o extension since emcc will output an object file, not a bitcode file')
+    if all(get_file_suffix(i) in ASSEMBLY_ENDINGS for i in options.input_files):
+      cmd = get_clang_command_asm() + newargs
     else:
-      cmd = get_clang_command() + inputs
-    if options.output_file:
-      cmd += ['-o', options.output_file]
-      if get_file_suffix(options.output_file) == '.bc' and not settings.LTO and '-emit-llvm' not in state.orig_args:
-        diagnostics.warning('emcc', '.bc output file suffix used without -flto or -emit-llvm.  Consider using .o extension since emcc will output an object file, not a bitcode file')
+      cmd = get_clang_command() + newargs
     shared.exec_process(cmd)
     assert False, 'exec_process does not return'
 
   # In COMPILE_AND_LINK we need to compile source files too, but we also need to
   # filter out the link flags
   assert state.mode == Mode.COMPILE_AND_LINK
   assert not options.dash_c
+  compile_args, input_files = split_linker_flags(options, state, newargs)
   compile_args = filter_out_link_flags(compile_args)
```

Comment:
As a follow up I hope to combine these two functions, but it makes this PR easier to follow if I leave them separate for now.